framework/src/base/FEProblem.C

```diff
@@ -704,16 +709,47 @@ FEProblem::getMaxScalarOrder() const
 }
 
 void
+FEProblem::checkNonlocalCoupling()
```

Comment:
This looks really expensive... even if it's only happening once per timestep... n_vars*n_subdomains could be HUGE for some problems (like neutronics).

Somehow you need to do this up front and cache the data...